### ISSUE TITLE
test(esp-idf): fake_esp_transport's static state is destroyed under a live recv_loop — TSan flake in esp_ws_client_dial

### ISSUE THREAD

**Issue by AvatarSD** [avatarsd-llc/libtracer#1456]
Surfaced on the `tsan (slot=hazard_slot_t)` leg of #1451 ([job](https://github.com/avatarsd-llc/libtracer/actions/runs/32406028875/job/96545209704)), 209/210. The sibling `tsan (slot=sp_atomic_slot_t)` leg passed the identical code, and `gh run rerun --failed` came back green — so it is timing-dependent, and it is **not** in libtracer proper. It is in the ESP-IDF host test double.

## The race

```
~state_t                      integrations/esp-idf/tests/host/fake_esp_transport.cpp:29
  (static destruction of the function-local `st()::s`, main thread, at process exit)
       vs
esp_transport_destroy         fake_esp_transport.cpp:328   — `st().live.erase(t)`
  tr::net::esp_ws_client_link_t::connect_once()   esp_ws_client_link.cpp:480
  tr::net::esp_ws_client_link_t::recv_loop()      esp_ws_client_link.cpp:827
  (the link's own thread, esp_ws_client_link.cpp:289)
```

TSan names the location precisely: `global '(anonymous namespace)::st()::s' of size 568`.

The fake keeps its whole state in a function-local static:

```cpp
struct state_t {
    std::mutex m;                      // guards everything below
    std::condition_variable cv;
    std::vector<fake_ws::frame_t> frames;
    ...
    std::map<esp_transport_item_t*, bool> live;
};

state_t& st() { static state_t s; return s; }
```

and `esp_transport_destroy` erases from `st().live` under `st().m`. The link's `recv_loop` thread outlives `main`, so at exit `~state_t` runs `~map` on `live` while that thread is inside `erase` — and `m`, the mutex the erase holds, is being destroyed in the same teardown. Locking cannot save it: the guard is part of the object going away.

Two distinct reports in the one run — `~_Rb_tree` reading `_M_mbegin()` against `_M_erase_aux` → `_M_reset()` — which is the same object seen from both sides.

## Why it is a real bug and not just TSan noise

Destroying a `std::map` and a `std::mutex` while another thread holds that mutex and mutates that map is undefined behaviour outright. It happens to survive most runs because the window is a few instructions wide at process exit. It will keep reddening the TSan legs at random, on PRs that have nothing to do with the WS client — exactly how it surfaced here, on an RFC-0014 enumeration change.

## Suggested fix

The fake needs a teardown the test drives, rather than relying on static destruction order:

1. **Join the link's thread before the fake's state dies.** `esp_ws_client_link_t`'s destructor should stop and join `recv_loop`; the test should destroy the link before returning from `main`. This is the real fix — a link that leaves its receive thread running past its own destruction is a lifetime hazard on the device too, not only in the fake.
2. **Failing that, give the fake an explicit lifetime.** Replace the function-local static with a `fake_ws::reset()` / scoped-fixture the test constructs and destroys while it still owns every thread, or leak the state deliberately (`static state_t* s = new state_t;`) so nothing is destroyed under a live reader. Leaking is the cheap unblock; (1) is the one worth doing.

Worth checking whether the other ESP-IDF host tests sharing `fake_esp_transport.cpp` have the same exposure.

Found by the merge train while landing #1451; not fixed there because it is unrelated to that change and deserves its own diagnosis.

**Comment by AvatarSD** [comment]
## Triage: confirmed real, and the repo already contains the fix pattern

Verdict: **`ready-for-agent`**. The race is real (not TSan noise), the mechanism is pinned end to end, and a sibling test file already implements the exact remedy — so this is a mechanical port, not a design call.

### What I verified

**The fake's state is a function-local static, mutated under its own mutex.**
- `integrations/esp-idf/tests/host/fake_esp_transport.cpp:29` — `struct state_t` (mutex `m`, `std::map<esp_transport_handle_t,bool> live`, condvars, vectors).
- `fake_esp_transport.cpp:95-98` — `state_t& st() { static state_t s; return s; }`.
- `fake_esp_transport.cpp:325-330` — `esp_transport_destroy` takes `st().m` and does `st().live.erase(t)`. Exactly the `~map`/`_M_erase_aux` pairing TSan named.

**The thread that reaches it at exit is a deliberately DETACHED orphan, not a stray.**
- `integrations/esp-idf/libtracer/esp_ws_client_link.cpp:338-351` — the destructor condemns an in-flight dial and calls `recv_thread_.detach()` instead of joining (`else if (... ) join()` only when no dial is in flight).
- `esp_ws_client_link.cpp:470-486` — the condemned orphan releases its own pair: `esp_transport_close(slot->ws)` / `esp_transport_destroy(slot->ws)` / `esp_transport_destroy(slot->tcp)` at **:480**, which is the exact frame in the reported stack.
- `esp_ws_client_link.cpp:63-83` — the detach is the documented `#1058` bound: the orphan may outlive the link by up to `kDialTimeoutMs` (half the watchdog window, 2.5 s by default).

**The dial test returns from `main` without waiting for that orphan.**
- `integrations/esp-idf/tests/host/esp_ws_client_dial_test.cpp:334-341` — `test_the_effective_config_is_readable()` constructs a link at :338 and destroys it at :341 with no liveness wait in between, so the recv thread is very likely still inside `esp_transport_connect` → condemned → detached. `main` (:357-369) then returns within microseconds of the remaining pure-arithmetic checks. Window: the orphan's `esp_transport_destroy` at link.cpp:480 versus `~state_t` on the main thread. That is the whole flake.

**The repo already states the rule and enforces it — in one file only.**
- `integrations/esp-idf/tests/host/esp_ws_client_teardown_test.cpp:165-177` — `drained_fake()` waits for `fake_ws::live_handles() == 0 && fake_ws::dialers_inside() == 0`, and its own doc comment says it outright: *"the orphaned recv thread is DETACHED, so a case that returns while it is still inside the transport leaves a running thread at process exit (a crash under ASan/TSan)."*
- The oracles it uses are already public: `fake_esp_transport.hpp:166` (`live_handles()`) and `:208` (`handle_misuse()`).
- `esp_ws_client_dial_test.cpp`, `esp_ws_client_recv_test.cpp`, `esp_ws_client_liveness_test.cpp`: **zero** occurrences of `live_handles` / `dialers_inside` / `drained`. All three share the same exposure the teardown suite designed around. (Answers the report's "worth checking whether the other host tests have the same exposure" — they do.)

### On the two suggested fixes

Fix (1) as written — "the destructor should stop and join `recv_loop`" — should **not** be taken. The non-joining destructor is the `#1058` ruling itself (`esp_ws_client_link.cpp:325-337`): `esp_transport_connect` takes no cancellation, so joining would charge the destroying task a full dial bound and put teardown at risk of the task watchdog. Reverting it is a design change requiring an RFC-grade argument, not a test fix. The lifetime hazard it names is real but is already bounded and owned by the link (the orphan touches no link member after `connect_once` resolves, `:483-485`).

Fix (2) is the right one, in its stronger form: **do not weaken the fake's lifetime — give the tests the drain the teardown suite already has.**

### Proposed bounded fix

1. Lift `drained_fake()` (or just the two-predicate wait) into a shared host-test helper, or copy it into each suite.
2. Call it at the end of every case in `esp_ws_client_dial_test.cpp`, `esp_ws_client_recv_test.cpp`, `esp_ws_client_liveness_test.cpp` that destroys a link — minimally `esp_ws_client_dial_test.cpp:341`, and defensively before `main` returns in all three.
3. Treat a drain timeout as a check failure, not a silent skip: an orphan still holding a pair after 5 s is a genuine release bug, and this is the same release oracle the teardown suite rests on.

No library change; no fake redesign; no `new state_t` leak (leaking would hide the release-oracle signal the suite depends on).

### Reproduction note

TSan binaries in this environment fault on ASLR — run TSan reproductions under `setarch $(uname -m) -R`. The window is a few instructions wide at process exit, so expect to need a loop of runs (or an artificial delay before `~state_t`) to hit it locally rather than a single reproduction.

### Sequencing

Independent of the Q8 wave (#1461/#1462) — it touches only `integrations/esp-idf/tests/host/`, so it can run at any time.
